fiftyone/brain/internal/core/leaky_splits.py

```diff
@@ -0,0 +1,497 @@
+import fiftyone.core.validation as fov
+import fiftyone.zoo as foz
+
+fbu = fou.lazy_import("fiftyone.brain.internal.core.utils")
+
+_DEFAULT_MODEL = "clip-vit-base32-torch"
+_DEFAULT_BATCH_SIZE = None
+
+
+def compute_leaky_splits(
+    samples,
+    brain_key,
+    split_views=None,
+    split_field=None,
+    split_tags=None,
+    threshold=0.2,
+    embeddings_field=None,
+    model=None,
+    model_kwargs=None,
+    patches_field=None,
+    metric="cosine",
+    **kwargs,
+):
+    """Uses embeddings to index the samples or their patches so that you can
+    find leaks.
+
+    Calling this method only creates the index. You can then call the methods
+    exposed on the retuned object to perform the following operations:
+
+    -   :meth:`leaks <fiftyone.brain.core.internal.leaky_splits.LeakySplitIndexInterface.leaks>`:
+        Returns a view of all leaks in the dataset.
+
+    -   :meth:`view_without_leaks <fiftyone.brain.core.internal.leaky_splits.LeakySplitIndexInterface.view_without_leaks>`:
+        Returns a subset of the given view without any leaks.
+
+    -   :meth:`tag_leaks <fiftyone.brain.core.internal.leaky_splits.LeakySplitIndexInterface.tag_leaks>`:
+        Tags leaks in the dataset as leaks.
+
+
+    Args:
+        samples: a :class:`fiftyone.core.collections.SampleCollection`
+        split_views (None): a list of :class:`fiftyone.core.view.DatasetView`
+            corresponding to different splits in the datset. Only one of
+            `split_views`, `split_field`, and `splits_tags` need to be used.
+        split_field (None): a string name of a field that holds the split of the sample.
+            Each unique value in the field will be treated as a split.
+            Only one of `split_views`, `split_field`, and `splits_tags` need to be used.
+        split_tags (None): a list of strings, tags corresponding to differents splits.
+            Only one of `split_views`, `split_field`, and `splits_tags` need to be used.
+        threshold (0.2): The threshold to run the algorithm with. Values between
+            0.1 - 0.25 tend to give good results.
+        patches_field (None): a sample field defining the image patches in each
+            sample that have been/will be embedded. Must be of type
+            :class:`fiftyone.core.labels.Detection`,
+            :class:`fiftyone.core.labels.Detections`,
+            :class:`fiftyone.core.labels.Polyline`, or
+            :class:`fiftyone.core.labels.Polylines`
+        embeddings_field (None): field for embeddings to feed the index. This argument's
+            behavior depends on whether a ``model`` is provided, as described
+            below.
+
+            If no ``model`` is provided, this argument specifies the field of pre-computed
+            embeddings to use:
+
+            If a ``model`` is provided, this argument specifies where to store
+            the model's embeddings:
+        brain_key (None): a brain key under which to store the results of this
```

Comment:
Given my comments below about how I don't think this method should store its own embeddings, I'm not sure it needs a `brain_key`?

A `brain_key` is only necessary/useful if a method permanently stores data on the samples or the dataset itself.

For context, `fob.compute_exact_duplicates()` does not use a brain key because it doesn't store anything; it just returns a moment in time snapshot of the duplicates that currently exist on the dataset. It is up to the user to do what they will with that information.

Reply:
Not a strongly held opinoin, but I disagree. The "leaks" will depend on the hashing method, and on the view you are computing on. I would expect results to differ when either or both is changed. One could absolutely be interested in comparing two runs.

Reply:
My confusion is in how this concretely looks. Suppose I ran two leaky split methods and closed my FO session. What does it look to start a new session, call `load_dataset()` and `load_brain_results()` twice and start comparing my previous results? Is anything actually cached? 

I could definitely be missing something, its just hard to reason about this without some example code showing how one works with the output of this method.

Reply:
I tend to agree with Jacob here, I definitely noticed a difference in results between different embedding models. Let me know how to proceed here.

Reply:
We allow `compute_visualization()` to store this type of information, and then in the panel we can toggle between runs. Could easily see a panel for leakage that allows you to toggle between leakage runs, loading the appropriate views and allowing you to take action on the results of one of these runs

Reply:
I don't disagree, I just need some example code in the PR description so I can see what this looks like and try it myself 😄

Reply:
```
model = foz.load_zoo_model("clip-vit-base32-torch")
dataset.compute_embeddings(model, "embeds_clip", batch_size=32, num_workers=16)
index_clip, leaks_clip = ls.compute_leaky_splits(dataset, "foobar", split_field='split', threshold=0.2, embeddings_field='embeds_clip')
```

Reply:
Not sure how to change the PR description after making it, sorry about that.